sim/firesim-lib/src/main/cc/bridges/dromajo.cc

```diff
@@ -0,0 +1,388 @@
+    int iaddr_width,
+    int insn_width,
+    int wdata_width,
+    int cause_width,
+    int tval_width,
+    int num_streams,
+    DROMAJOBRIDGEMODULE_struct * mmio_addrs,
+    long dma_addr) : bridge_driver_t(sim)
+{
+    // setup max constants given from the RTL
+    this->_num_streams = num_streams;
+
+    this->_valid_width = 1;
+    this->_iaddr_width = iaddr_width/8;
+    this->_insn_width = insn_width/8;
+    this->_wdata_width = wdata_width/8;
+    this->_priv_width = 1;
+    this->_exception_width = 1;
+    this->_interrupt_width = 1;
+    this->_cause_width = cause_width/8;
+    this->_tval_width = tval_width/8;
+    this->_valid_offset = 0;
+    this->_iaddr_offset = this->_valid_offset + this->_valid_width;;
+    this->_insn_offset = this->_iaddr_offset + this->_iaddr_width;
+    this->_wdata_offset = this->_insn_offset + this->_insn_width;
+    this->_priv_offset = this->_wdata_offset + this->_wdata_width;
+    this->_exception_offset = this->_priv_offset + this->_priv_width;
+    this->_interrupt_offset = this->_exception_offset + this->_exception_width;
+    this->_cause_offset = this->_interrupt_offset + this->_interrupt_width;
+    this->_tval_offset = this->_cause_offset + this->_cause_width;
+
+    // setup misc. state variables
+    this->_mmio_addrs = mmio_addrs;
+    this->_dma_addr = dma_addr;
+    this->_stream_idx = 0;
+    this->dromajo_failed = false;
+    this->dromajo_exit_code = 0;
+    this->dromajo_state = NULL;
+    this->dromajo_cosim = true; // by default enable cosim
+    this->saw_int_excp = false; // used to not trigger interrupts multiple times
+
+    // setup dromajo variables
+    std::string dromajo_dtb_arg = std::string("+drj_dtb=");
+    std::string dromajo_bin_arg = std::string("+drj_bin=");
+    std::string dromajo_rom_arg = std::string("+drj_rom=");
+    bool dtb_found = false;
+    bool bin_found = false;
+    bool bootrom_found = false;
+    this->dromajo_dtb = "";
+    this->dromajo_bootrom = "";
+    this->dromajo_bin = "";
+    const char *d_rom = NULL;
+    const char *d_dtb = NULL;
+    const char *d_bin = NULL;
+
+    for (auto &arg: args) {
+        if (arg.find(dromajo_dtb_arg) == 0) {
+            d_dtb = const_cast<char*>(arg.c_str()) + dromajo_dtb_arg.length();
+            this->dromajo_dtb = std::string(d_dtb);
+            dtb_found = true;
+        }
+        if (arg.find(dromajo_rom_arg) == 0) {
+            d_rom = const_cast<char*>(arg.c_str()) + dromajo_rom_arg.length();
+            this->dromajo_bootrom = std::string(d_rom);
+            bootrom_found = true;
+        }
+        if (arg.find(dromajo_bin_arg) == 0) {
+            d_bin = const_cast<char*>(arg.c_str()) + dromajo_bin_arg.length();
+            this->dromajo_bin = std::string(d_bin);
+            bin_found = true;
+        }
+    }
+
+    if (!dtb_found || !bin_found || !bootrom_found) {
+        printf("[WARNING] Missing Dromajo input file(s) (make sure you have a dtb, bin, and bootrom passed in)\n");
```

Comment:
IIRC, all of the output ends up in the `uartlog` (including the Dromajo output). However, I will double-check this when I rebuild and AFI with the changes.

Reply:
Just checked the `uartlog` output... The `printf`s get piped to the `uartlog`. Same for `stderr` output coming from Dromajo. Like this:

```
Script started on Fri 15 May 2020 06:38:19 PM UTC
AFI PCI  Vendor ID: 0x1d0f, Device ID 0xf000^M
Using xdma write queue: /dev/xdma0_h2c_0^M
Using xdma read queue: /dev/xdma0_c2h_0^M
UART0 is here (stdin/stdout).^M
command line for program 0. argc=56:^M
+permissive +mm_relaxFunctionalModel_0=0 +mm_openPagePolicy_0=1 +mm_backendLatency_0=2 +mm_sched
TraceRV 0: Tracing disabled, since +tracefile was not provided.^M
random min: 0x0, random max: 0xffffffffffffffff^M
[INFO] Dromajo: Attached Dromajo to 3 instruction traces^M
[INFO] Dromajo command: ^M
./dromajo --compact_bootrom --custom_extension --clear_ids --reset_vector 0x10040 --bootrom br-b
Commencing simulation.^M
3 0x0000000000010040 (0x00000517) x10 0x0000000000010040 DASM(0x00000517)^M
3 0x0000000000010044 (0xfc050513) x10 0x0000000000010000 DASM(0xfc050513)
```